### ISSUE TITLE
After firefox crash, xowa doesn't work (minor)

### ISSUE THREAD

**Issue by pirate486743186** [piotrex/XOWA-Addon#5]
A minor bug.

When there is a crash in Firefox, the old xowa stays alive, while Firefox opens a second xowa.
But it doesn't work, you just get a blank screen, no error messages.
If you kill the old by hand, it still doesn't work.
You have to kill the old xowa and restart Firefox.

**Comment by gnosygnu** [comment]
Thanks for the report. I think the problem is that XOWA opens up a port on 55000. When Firefox crashes, XOWA is still running, but still keeps the port open on 55000. The next time the addon launches, it creates and uses a new XOWA. However, this new XOWA fails because 55000 is still kept open by the first XOWA.

The fix here is to check if there are any open instances of XOWA and try to reuse them. Unfortunately, I don't know an easy way to do it. I'll put it down on my todo list, but it may be quite a while.

**Comment by pirate486743186** [comment]
You can't make xowa die when Firefox dies?
also zombie xowa is of an important size, it should die along with Firefox.

Also, even if i killed the old xowa manually, new xowa still doesn't work.

**Comment by gnosygnu** [comment]
> You can't make xowa die when Firefox dies?

It's hard. When Firefox crashes, it crashes before it can do anything. When a new Firefox starts, it doesn't have any knowledge of the old Firefox. In particular, it doesn't know that the background XOWA (java) process was created by it. So it creates a new XOWA in its place, but runs into problems because the old XOWA is still holding on to the socket port

I made a few tries tonight, but couldn't come up with any way for the new Firefox to claim the old Firefox's XOWA process.

> also zombie xowa is of an important size, it should die along with Firefox.

Agreed, but again, Firefox crashes before anything can be done.

> Also, even if i killed the old xowa manually, new xowa still doesn't work.

If you kill the old XOWA after the old Firefox dies, but before you start a new Firefox, it should work.

Sorry, but this is admittedly not my expertise. @piotrex is probably much better at this.

**Comment by pirate486743186** [comment]
I was thinking, to have xowa it self kill it self if it detects Firefox died?
xowa is quite big, and should somehow go away even if Firefox crashes.
I bet java has some neat way to deal with this.

**Comment by gnosygnu** [comment]
> I bet java has some neat way to deal with this.

Java has a socket timeout mechanism. I'm not sure if that will help as I'd have to set it to a large number (30 min). Otherwise, it might time out while reading.

**Comment by pirate486743186** [comment]
What if it manually checked, that The original Firefox process still exist? The PID?
Say, every 10seconds?

**Comment by gnosygnu** [comment]
Unfortunately, not an easy way. The PID would be OS-dependent (Windows vs Linux vs Mac OS X). There isn't any standard Java code for it, though JNA is possible. Also checking PID may still not work (someone could have 2 Firefox's open)

**Comment by pirate486743186** [comment]
Hmmm
What about instead, have the addon display a warning message with instruction for the user what to do?

**Comment by gnosygnu** [comment]
Sorry for the late reply.

I think the warning message would need to show every time they launch the add-on which could be annoying. There's no way to show the warning message when the addon just died.

Let me know if I misunderstood. Thanks.

**Comment by pirate486743186** [comment]
Not every time the addon launches. It can't detect that it wasn't able to connect with xowa?
You could make a small change in both the addon and xowa, say a handshake. 
"successful docking" :P
If the handshake fails, then bam the warning.
I was thinking a warning, inside the main window of Firefox, not a pop up.

Say they ping each other from time to time, so they can detect if the other is still there, and take appropriate action (and forget the warning).

**Comment by gnosygnu** [comment]
> It can't detect that it wasn't able to connect with xowa?

Nope. The addon always launches a new instance if it can't find one. It doesn't try to connect to any existing ones.

> You could make a small change in both the addon and xowa, say a handshake. 

Yeah, I tried to do something like this a few nights ago. Basically, I tried to get the addon to see if XOWA was already running. Wasn't successful. :(

> Say they ping each other from time to time, so they can detect if the other is still there, and take appropriate action

This is actually sort of hard. XOWA is running as a server, so technically it is just serving requests to clients. In theory, multiple clients can connect, and the server ordinarily doesn't keep track of them, as a client can be a one-time visitor.

I'll take another crack at detecting it from the addon side, but again, the addon code is just not my area of expertise...

**Comment by pirate486743186** [comment]
beh....
What about xowa the server having a panel applet.
This way, there is something visible, when xowa the server is active.
You could also control it from there, and maybe display something useful?

**Comment by gnosygnu** [comment]
Honestly, I think that would be much harder than anything above.

Sorry, no easy fixes (at least none that I know of).

**Comment by pirate486743186** [comment]
oh well...
